### PR TITLE
Digest Service - 2 Query Version

### PR DESCRIPTION
**Pull request by jon-betts**
For:

 * https://github.com/hypothesis/lms/issues/5194

This is a refactor of the digest service which mainly aimed to reduce the number of queries executed.

 * Previously: **5** different queries, **executed multiple times per user / course**
 * Now: **2** different queries, **executed once each** regardless of the number of users / courses

## Aims of the new design

In order to facilitate this approach, the service had to be rewritten. The current design has a few aims:

####  All DB access is in the helper class

* No DB access is performed in the main service at all
* It's purely responsible for data manipulation and calling other services
* This helps a lot with separating responsibilities and makes mocking in the tests easier
* This keeps the really gnarly fixtures in one place
   
#### The various parts of the design are less interconnected

* Previously the user look-up was performing multiple functions and associated with the later lookups for the courses
* Now they are completely separate and don't interact
* There are fewer simpler data structures at most stages of the process
* The process is more like a pipe-line and you can concentrate on one part without worrying about the whole as much
* Each component is now in it's own file with it's own tests

#### Less active models

  * All the models used in this design are pure data-classes
  * They only have a few simple methods to present their data

#### Less overlapping responsibilities

 * I've aimed to have each part be doing fewer things
 * And also to try to be the only place responsible for that thing
 * e.g.
     * The `DigestAssistant` does 100% of the DB interaction
        * The two methods on it are 100% responsible for generating and assembling users / courses
        * Those objects aren't fundamentally changed from then on (just moved about and arranged)
        * The `get_h_users` method is 100% about getting display names and emails, and nothing to do with teacher lookups etc.
     * The `DigestService` has a `get_digests` method which is responsible for creating `Digest` objects
        * It doesn't format them
        * It doesn't email them or choose whether to email them
    * The `Digest` object is responsible for serializing itself

## Trade offs

<table>
    <tr>
        <th>Old design</th>
        <th>New design</th>
    </tr>
    <tr>
        <td>✅ Simple queries (5) </td>
        <td>Complex queries (2) </td>
    </tr>
    <tr>
        <td>Many queries executed per item</td>
        <td>✅ Two queries executed for all</td>
    </tr>
    <tr>
        <td>🤷 More manipulation in Python</td>
        <td>🤷 More manipulation in Postgres</td>
    </tr>
    <tr>
        <td>🤷 Uses our ORM objects</td>
        <td>🤷 Uses dataclasses</td>
    </tr>
    <tr>
        <td>More complex tests</td>
        <td>✅ Simpler tests</td>
    </tr>
    <tr>
        <td>More code (755 code / 78 comments)</td>
        <td>✅ Less code* (452 code / 79 comments)</td>
    </tr>
</table>

_* This is mostly due to the smaller tests. The service code is of similar size (~210 lines), although much more of it is now a big SQL query (73 lines). So the amount of Python proper has dropped._

### CODE REVIEW

File: lms/services/digest/_models.py [hunk 1]
```diff
@@ -0,0 +1,76 @@
...
+            if course.learner_annotations:
+                digest["total_annotations"] += len(course.learner_annotations)
+                digest["courses"].append(
+                    {
+                        "title": course.title,
+                        "num_annotations": len(course.learner_annotations),
+                    }
+                )
+
+        return digest
```
Comment (jon-betts) [inline lms/services/digest/_models.py:76]: I like and dislike having this here.

Having this be separate from the service means we can test these bits very easily without having to kick off (and therefore fixture) a whole pile of stuff to inspect the output.

The downside is it means what we are actually sending to the email service is a function call farther away. We could do this inline.

File: lms/services/digest/_models.py [hunk 2]
```diff
@@ -0,0 +1,76 @@
...
+    """The course's authority provided id."""
+
+    aka: List[str]
+    """All authority provided ids which map to the course from groups etc."""
+
+    instructors: List[str]
+    """A list of H userids for users who are instructors for the course."""
+
+    annotations: List[dict] = None
+    """A list of annotations in the course."""
```
Comment (jon-betts) [inline lms/services/digest/_models.py:37]: This PR would be more compact with less doc-strings, but we don't actually get any prizes for few lines if they are there for visual clarity.

I think spacing things out enough can eventually become a readability issue, but this makes it easier to associate the comment to the attr I think. If you have them all squished, or not documented, it's a bit less clear what things are.

We can try to get around that with naming, but `list_of_instructor_h_userids` or `authority_provided_id_for_self_and_children` etc. kind of suck.

File: lms/services/digest/service.py [hunk 3]
```diff
@@ -0,0 +1,109 @@
...
+        """
+        Get activity digests for users over a period.
+
+        :param audience: List of H userids to send digests to
+        :param updated_after: Search for activity after this date
+        :param updated_before: Search for activity before this date
+        """
+        annotations = self._h_api.get_annotations(
+            audience, updated_after, updated_before
+        )
```
Comment (jon-betts) [inline lms/services/digest/service.py:70]: This is kind of annoying, as it bloats our interface a lot. We could accept a list of annotations instead in this and the send email method.

This would mean our tests wouldn't have this sort of thing everywhere:

```
        svc.send_instructor_email_digests(
            audience, sentinel.updated_after, sentinel.updated_before
        )

        h_api.get_annotations.assert_called_once_with(
            audience, sentinel.updated_after, sentinel.updated_before
        )
```

This would complicate the caller a bit, but we only have one caller...

File: lms/services/digest/service.py [hunk 4]
```diff
@@ -0,0 +1,109 @@
...
+                instructors_to_courses[h_userid].append(course)
+
+        for audience_user in self._assistant.get_h_users(h_userids=audience):
+            yield Digest(
+                audience_user=audience_user,
+                courses=instructors_to_courses.get(audience_user.h_userid, []),
+            )
+
+    @classmethod
+    def _add_annotations_to_courses(cls, courses, annotations):
```
Comment (jon-betts) [inline lms/services/digest/service.py:96]: No particular reason to put this in something separate, but it's not clear if you do it in a line that these two things are are only being done for each other.

File: lms/services/digest/service.py [hunk 5]
```diff
@@ -0,0 +1,109 @@
...
+                for annotation in annotations
+            }
+        )
+        self._add_annotations_to_courses(courses, annotations)
+
+        # Make a mapping from instructors to courses
+        instructors_to_courses = defaultdict(list)
+        for course in courses:
+            for h_userid in course.instructors:
+                instructors_to_courses[h_userid].append(course)
```
Comment (jon-betts) [inline lms/services/digest/service.py:87]: We do this type of thing twice. It seems like the kind of thing we would be able to accomplish better, but I'm not sure how.

File: lms/services/digest/_digest_assistant.py [hunk 6]
```diff
@@ -0,0 +1,126 @@
...
+        id, or have children with the given authority provided id.
+
+        This will return a list of course object which includes other authority
+        provided ids which might link to this group (from sections etc.) and
+        a list of H userids for instructors in the course.
+
+        :param authority_provided_ids: A list of authority provided ids to
+            start the search.
+        """
+
```
Comment (jon-betts) [inline lms/services/digest/_digest_assistant.py:63]: This is a belter and no mistake.

I don't think SQLAlchemy makes this any easier on us here. I find the raw SQL considerably easier to read. However it is convenient in a few places, as we can directly refer to things we would have to use binds for otherwise.

In any case, it does show that we can do fairly complicated SQL in SQLAlchemy. It does require me to search the internet for nearly every part of it though, which I dislike. It's often pretty hard to know where to start with something like:

```(ARRAY_AGG(col ORDER BY foo) FILTER (WHERE thing)```

Half of that is in the main library, some is Postgres specific, and there's no rhyme or reason to the syntax required.

File: lms/services/digest/_models.py [hunk 7]
```diff
@@ -0,0 +1,76 @@
...
+    display_name: str
+    """A human readable display name."""
+
+    email: str
+    """Email for the user."""
+
+
+@dataclass
+class HCourse:
+    """A class representing courses as groups in H."""
```
Comment (jon-betts) [inline lms/services/digest/_models.py:22]: This is a bit of a fib, but I think it gets the idea across. We are talking about a group in H not a grouping in LMS.

File: lms/services/digest/_models.py [hunk 8]
```diff
@@ -0,0 +1,76 @@
...
+            if annotation["author"]["userid"] not in self.instructors
+        ]
+
+
+@dataclass
+class Digest:
+    """A digest of activity for a given user."""
+
+    audience_user: HUser
+    """The user who is the intended recipient of the information."""
```
Comment (jon-betts) [inline lms/services/digest/_models.py:55]: Technically the audience object isn't required for the serialisation. We could omit this and pass it separately.

Interestingly at the moment our digest contents aren't dependent on the user at all, only the courses.

If this was expensive we could separate the two and do the digest serialization separately and cache it. However it's cheap as chips and this is easier.

File: lms/services/digest/service.py [hunk 9]
```diff
@@ -0,0 +1,109 @@
+from collections import defaultdict
+from datetime import datetime
+from typing import Iterator, List
+
+from lms.services.digest._digest_assistant import DigestAssistant
+from lms.services.digest._models import Digest
+from lms.services.h_api import HAPI
+from lms.services.mailchimp import EmailRecipient, EmailSender, MailchimpService
```
Comment (jon-betts) [inline lms/services/digest/service.py:8]: `EmailTo, EmailFrom` ?

File: lms/tasks/email_digests.py [hunk 10]
```diff
@@ -78,24 +78,23 @@ def send_instructor_email_digests(
...
-                updated_after,
-                updated_before,
+            request.find_service(DigestService).send_emails(
+                audience=h_userids,
+                # Caution: `datetime.fromisoformat()` doesn't support all ISO
+                # 8601 strings! This only works for the subset of ISO 8601
+                # produced by `datetime.isoformat()`.
+                updated_after=datetime.fromisoformat(updated_after),
+                updated_before=datetime.fromisoformat(updated_before),
                 override_to_email=override_to_email,
```
Comment (jon-betts) [inline lms/tasks/email_digests.py:99]: The functional change here is because I renamed `send_instructor_email_digests` -> `send_emails`.

The rest was just some tidying while I was here.

File: tests/unit/lms/services/digest/_digest_assistant_test.py [hunk 11]
```diff
@@ -0,0 +1,89 @@
...
+                # Check we can get more than one at a time
+                ("h_userid_2", "email_2", "display_2"),
+            )
+        ):
+            factories.User(
+                h_userid=userid,
+                email=email,
+                display_name=display_name,
+                # Ensure the insert order is the updated order too
+                updated=datetime.now() + timedelta(days=days),
```
Comment (jon-betts) [inline tests/unit/lms/services/digest/_digest_assistant_test.py:29]: I don't think this is required, as the insert order happens to do this for us, but this seems belt and braces? 

I can remove it if you don't think it's required.

File: tests/unit/lms/services/digest/service_test.py [hunk 12]
```diff
@@ -0,0 +1,150 @@
...
+            mailchimp_service.send_template.assert_not_called()
+
+    def test_send_email_requires_some_annotations(self, svc, digest, mailchimp_service):
+        digest.courses[0].annotations = []
+
+        svc.send_emails(sentinel.audience, sentinel.after, sentinel.before)
+
+        mailchimp_service.send_template.assert_not_called()
+
+    def test_get_digests(self, svc, h_api, digest_assistant):
```
Comment (jon-betts) [inline tests/unit/lms/services/digest/service_test.py:65]: We aren't testing lots of different combinations of things, but I'm not sure how many it's worth testing. A lot of the edge casey things we were testing here before are now someone else's responsiblity.

As long as the annotations get on the right course and assigned to the user it's all gravy.

The two ways to go with this I think are:

 * Add different tests which cover different aspects, but leaves us with a horrible fixturing problem
 * Add some more elements to this test to cover more cases

File: lms/services/digest/_models.py [hunk 13]
```diff
@@ -0,0 +1,76 @@
...
+    """A class representing H users."""
+
+    h_userid: str
+    """The H userid like 'acct:name@lms.example.com'."""
+
+    display_name: str
+    """A human readable display name."""
+
+    email: str
+    """Email for the user."""
```
Comment (jon-betts) [inline lms/services/digest/_models.py:17]: Having all these fields mandatory feels correct for the actual code, but sucks for the tests. We often have to fill out loads of fields we don't care about to get a fixture.

I'm not sure which is worse?

File: lms/services/digest/_models.py [hunk 14]
```diff
@@ -0,0 +1,76 @@
...
+
+    @cached_property
+    def learner_annotations(self):
+        """Get the annotations, but filtered to users who are learners."""
+
+        return [
+            annotation
+            for annotation in self.annotations
+            if annotation["author"]["userid"] not in self.instructors
+        ]
```
Comment (jon-betts) [inline lms/services/digest/_models.py:47]: Previously this was part of the code which generated the courses in the first place, but it doesn't really seem like it's anyone's business but our own.

This keeps the outside code simpler, but it also means that should we want to know the total number of annotations as well as the number from students, we can. In general I think it's more flexible to store all the info, and then whittle it down for display.

File: lms/services/digest/service.py [hunk 15]
```diff
@@ -0,0 +1,112 @@
...
+        self._h_api = h_api
+        self._mailchimp_service = mailchimp_service
+        self._sender = sender
+
+    def send_emails(
+        self,
+        audience: List[str],
+        updated_after: datetime,
+        updated_before: datetime,
+        override_to_email: str = None,
```
Comment (marcospri) [inline lms/services/digest/service.py:31]: Optional[str] if you don't want mypy and friends to yell at you.

File: lms/services/digest/_models.py [hunk 16]
```diff
@@ -0,0 +1,72 @@
...
+    authority_provided_id: str
+    """The course's authority provided id."""
+
+    aka: List[str]
+    """All authority provided ids which map to the course from groups etc."""
+
+    instructors: List[str]
+    """A list of H userids for users who are instructors for the course."""
+
+    annotations: List[dict] = None
```
Comment (marcospri) [inline lms/services/digest/_models.py:32]: If we were payed by number of data classes I'd make this one. Is hard to spot the the shape of these without reading multiple bits.

h_api.get_annotations could return that direclty.

**Review by marcospri — APPROVED** [review]
We have looked at this code together on a call and talk about it enough. I don't have any thing of substance to comment here, I read it again and got a good understanding of it.

I reckon it's easy to grok and it's more performant. I don't have strong opinions on how it compares structurally with the previous version. I do think that changes in the data flow -ie number of queries- warrants what it looks like a rewrite.

I would wait to merge this until next week and have a chat about the PR itself with @seanh  and next steps for this project, are we planning to start sending hundred of emails as soon as we merge this?

**Comment by seanh** [comment]
Just a note to say I'm looking into this, apologies for the delay

**Comment by seanh** [comment]
Apologies for the delay in getting back to this. [Noted in Slack already](https://hypothes-is.slack.com/archives/C1MA4E9B9/p1681131840814179) but I think we should close this PR: it has been very helpful in providing the DB queries that the ones in https://github.com/hypothesis/lms/pull/5234 were based on but I don't think it's gonna be helpful to merge a big rewrite at this time.
